# Natural Language Processing

## Importing the libraries

In [0]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Importing the dataset

In [0]:
dataset = pd.read_csv('Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3) # quoting = 3 is used to ignore double quotes

## Cleaning the texts

In [4]:
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
corpus = []
for i in range(0, 1000):
  review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])
  review = review.lower()
  review = review.split()

  ps = PorterStemmer()

  all_stopwords = stopwords.words('english')
  all_stopwords.remove('not')

  review = [ps.stem(word) for word in review if not word in set(all_stopwords)]
  review = ' '.join(review)
  corpus.append(review)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [5]:
print(corpus)

['wow love place', 'crust not good', 'not tasti textur nasti', 'stop late may bank holiday rick steve recommend love', 'select menu great price', 'get angri want damn pho', 'honeslti tast fresh', 'potato like rubber could tell made ahead time kept warmer', 'fri great', 'great touch', 'servic prompt', 'would not go back', 'cashier care ever say still end wayyy overpr', 'tri cape cod ravoli chicken cranberri mmmm', 'disgust pretti sure human hair', 'shock sign indic cash', 'highli recommend', 'waitress littl slow servic', 'place not worth time let alon vega', 'not like', 'burritto blah', 'food amaz', 'servic also cute', 'could care less interior beauti', 'perform', 'right red velvet cake ohhh stuff good', 'never brought salad ask', 'hole wall great mexican street taco friendli staff', 'took hour get food tabl restaur food luke warm sever run around like total overwhelm', 'worst salmon sashimi', 'also combo like burger fri beer decent deal', 'like final blow', 'found place accid could not

## Creating the Bag of Words model

In [0]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500)
X = cv.fit_transform(corpus).toarray()
y = dataset.iloc[:, -1].values

## Splitting the dataset into the Training set and Test set

In [0]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

## Training the Naive Bayes model on the Training set

In [0]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

## Predicting the Test set results

In [0]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 0]
 [1 0]
 [1 0]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [1 0]
 [1 0]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [1 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [1 0]
 [1 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 0]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 0]
 [0 1]
 [1 1]
 [1 1]
 [1 0]
 [0 1]
 [1 0]
 [1 1]
 [1 1]
 [0 0]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 1]
 [1 1]
 [1 1]

## Making the Confusion Matrix

In [0]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[55 42]
 [12 91]]


0.73

---
# Study Notes: Setting Up an NLP Sentiment-Analysis Project

## Lesson objective

In this project, we build a **natural language processing (NLP)** model for **sentiment analysis**. The input is a restaurant review, and the target is a binary label indicating whether the review is negative (`0`) or positive (`1`).

The workflow begins with the usual data-processing template, but text data requires additional preparation. We therefore import the dataset first and postpone the creation of the feature matrix `X` and target vector `y` until the text has been cleaned and converted into numerical features.

## 1. Importing the libraries

```python
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
```

These standard libraries are loaded at the start so they are available when needed. Not every imported library must necessarily be used in every stage of the notebook.

## 2. Why `X` and `y` are not created immediately

In a conventional tabular machine-learning problem, we often create `X` and `y` immediately after loading the dataset. Raw text, however, cannot be passed directly to most classical machine-learning algorithms. The reviews must first be:

1. cleaned and normalized;
2. tokenized and reduced to useful words; and
3. transformed into numerical features using a representation such as the **Bag-of-Words model**.

Only after vectorization do we create:

- `X`: the matrix of numerical text features; and
- `y`: the vector of sentiment labels.

## 3. Importing the TSV dataset

```python
dataset = pd.read_csv(
    'Restaurant_Reviews.tsv',
    delimiter='\t',
    quoting=3
)
```

Although the function is named `read_csv`, pandas can read many delimited text formats. The important arguments here are:

- **File name:** `Restaurant_Reviews.tsv` identifies the dataset.
- **`delimiter='\t'`:** specifies that columns are separated by tab characters rather than commas. `\t` is Python's escape sequence for a tab.
- **`quoting=3`:** tells the parser to treat quote characters as ordinary text instead of using them to group fields. The value `3` corresponds to `csv.QUOTE_NONE`.

> A more readable alternative is to import `csv` and write `quoting=csv.QUOTE_NONE` instead of using the numeric constant `3`. This setting is useful for this course dataset, but it is not a universal requirement for all text datasets.

## 4. Expected dataset structure

The dataset contains two columns:

| Column | Meaning | Example |
|---|---|---|
| `Review` | The restaurant-review text | `The food was excellent.` |
| `Liked` | Binary sentiment label | `1` for positive, `0` for negative |

This is a **supervised binary-classification** problem because each training review has a known label and there are two possible target classes.

## 5. Quick validation after loading

After importing the file, it is good practice to inspect it before continuing:

```python
dataset.head()
dataset.shape
dataset.info()
dataset['Liked'].value_counts()
```

These checks confirm that the columns were separated correctly, show the number of rows and columns, reveal missing values or unexpected data types, and display the class distribution.

## 6. Common pitfalls

- Use `delimiter='\t'`, not `delimiter='\nT'` or `delimiter='NT'`.
- Ensure the filename and working directory are correct; otherwise, Python raises `FileNotFoundError`.
- Do not build `X` directly from raw review strings when using a classical classifier; vectorize the cleaned text first.
- Do not assume `quoting=3` is always necessary. Choose parsing options to match the actual file format.
- Inspect a few rows after import so that delimiter or quoting problems are caught early.

## 7. Pipeline overview

`TSV file` → `pandas DataFrame` → `text cleaning` → `Bag of Words` → `X and y` → `train/test split` → `classifier` → `evaluation`

## Key takeaways

- Sentiment analysis predicts an opinion or emotional polarity from text.
- The restaurant-review task uses binary labels: negative (`0`) and positive (`1`).
- A TSV file uses tabs as separators, so it is loaded with `delimiter='\t'`.
- Raw text must be cleaned and converted into numbers before classical machine-learning models can use it.
- The feature matrix and target vector are created after text vectorization, not immediately after loading the dataset.

## Review questions

1. Why can `pd.read_csv()` read a `.tsv` file?
2. What does the escape sequence `\t` represent?
3. Why is the feature matrix `X` created only after the reviews are cleaned and vectorized?
4. What do the labels `0` and `1` represent in this dataset?
5. What is the more descriptive equivalent of `quoting=3`?


---
# Study Notes: Cleaning Text for Sentiment Analysis

## Lesson objective

Before a classical machine-learning model can learn from restaurant reviews, the raw text must be converted into a more consistent form. This lesson prepares each review by removing unwanted characters, converting text to lowercase, and splitting it into words. It also introduces stop-word removal and stemming, which are applied in the next part of the cleaning pipeline.

Text cleaning reduces irrelevant variation in the vocabulary. For example, `Food`, `food`, and `food!` should generally be recognized as the same term rather than treated as three unrelated features.

## 1. Importing the text-processing tools

```python
import re
import nltk

nltk.download('stopwords')

from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
```

### What each import provides

- **`re`** is Python's regular-expression module. It supports pattern-based searching and replacement in strings.
- **`nltk`** is the Natural Language Toolkit, a package containing linguistic datasets and text-processing utilities.
- **`nltk.download('stopwords')`** downloads NLTK's stop-word data if it is not already available in the current environment.
- **`stopwords`** gives access to language-specific lists of frequently occurring words.
- **`PorterStemmer`** implements the Porter stemming algorithm.

> The download usually needs to run only once per environment. In a controlled or production environment, NLTK data is commonly installed during setup rather than downloaded whenever the program runs.

## 2. Stop words

Stop words are common words—such as `the`, `a`, and `and`—that often contribute little to a simple Bag-of-Words classifier. Removing them can reduce the vocabulary size and eliminate noise.

However, stop-word removal must match the task. Negations such as `not`, `no`, and `never` can reverse sentiment, so removing them may discard important information. The implementation in this notebook deliberately keeps `not`:

```python
set(stopwords.words('english')) - {'not'}
```

Examples:

- `The meal was good` → `meal good`
- `The meal was not good` → `meal not good`

Keeping `not` helps preserve the difference between those two reviews, although a unigram Bag-of-Words model still does not fully represent word order.

## 3. Stemming

Stemming reduces related word forms to a shared stem. This can prevent the feature matrix from having separate columns for closely related terms.

```python
ps = PorterStemmer()
ps.stem('loved')  # typically returns 'love'
```

Stemming is a rule-based reduction process; it does **not** simply convert verbs to the present tense, and its output is not guaranteed to be a valid dictionary word. For example, a stemmer may reduce several related forms to an artificial-looking stem. The goal is consistent feature representation rather than perfect grammar.

A related technique, **lemmatization**, uses linguistic information to return dictionary base forms. It is often more precise but usually requires more processing.

## 4. Creating the corpus

```python
corpus = []
```

The corpus is a list that will store one cleaned string for each original review. The cleaning operation is applied review by review, and each completed result is appended to this list. The later vectorizer expects a collection of text documents, so this structure is suitable for building the Bag-of-Words model.

## 5. Iterating over the reviews

The course implementation uses:

```python
for i in range(0, 1000):
    ...
```

This works because the supplied dataset contains exactly 1,000 reviews, indexed from `0` through `999`. A more reusable version avoids hard-coding the number of rows:

```python
for i in range(len(dataset)):
    ...
```

Alternatively, the review strings can be iterated over directly:

```python
for raw_review in dataset['Review']:
    ...
```

## 6. Step 1—removing non-letter characters

```python
review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])
```

`re.sub(pattern, replacement, text)` finds every match of a pattern and replaces it with the specified value. Here:

- `[a-zA-Z]` represents one lowercase or uppercase English letter;
- `^` at the beginning of a character class means **not**; therefore, `[^a-zA-Z]` matches any character that is not an English letter;
- `' '` replaces every matched character with a space.

Replacing unwanted characters with spaces is important. Replacing them with an empty string could accidentally join separate words—for example, `good-food` could become `goodfood`.

> The meaning of `^` depends on its position in a regular expression. At the start of `[...]` it negates the character class; outside a character class it often means “start of the string.” It is not a universal “not” operator in Python.

This particular pattern removes punctuation, digits, emoji, accented characters, and all other non-ASCII letters. That is acceptable for this English course dataset, but multilingual or production NLP systems often use Unicode-aware tokenization instead.

## 7. Step 2—converting to lowercase

```python
review = review.lower()
```

Lowercasing normalizes capitalization so that terms such as `Food`, `FOOD`, and `food` map to the same vocabulary item. This usually reduces the number of features in a Bag-of-Words representation.

## 8. Step 3—splitting the review into words

```python
review = review.split()
```

Calling `split()` without an argument separates the string on whitespace and returns a list of tokens. For example:

```python
'the food was excellent'.split()
# ['the', 'food', 'was', 'excellent']
```

Tokenization is required because stop-word filtering and stemming are performed on individual words.

## 9. Cleaning sequence so far

```python
corpus = []

for i in range(len(dataset)):
    review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])
    review = review.lower()
    review = review.split()
```

At this point, each review is a list of lowercase tokens containing only the English letters matched by the regular expression. The next operations are to remove unwanted stop words, stem the remaining tokens, join them back into a string, and append the result to `corpus`.

## 10. Why cleaning affects the feature matrix

In a Bag-of-Words model, each selected vocabulary term becomes a feature column. Normalization can merge redundant variants and reduce dimensionality:

`Loved` + `loved!` + `love` → lowercase and punctuation removal → `loved` + `loved` + `love` → stemming → `love`

This produces a smaller and more consistent vocabulary. Nevertheless, more cleaning is not always better: removing sentiment-bearing words or useful distinctions can reduce predictive performance. Preprocessing choices should ultimately be evaluated on validation data.

## Common pitfalls

- Writing `review.lower` without parentheses stores a method reference instead of calling the method; use `review.lower()`.
- Using `range(1000)` makes the code dependent on this exact dataset size.
- Removing every stop word can unintentionally remove important negations.
- Assuming stemming always returns a grammatical base word leads to incorrect interpretations.
- Removing non-letters with an ASCII-only pattern may discard valuable information in other datasets.
- Replacing punctuation with `''` instead of `' '` can merge adjacent words.

## Key takeaways

- Text normalization makes vocabulary features more consistent.
- `re.sub('[^a-zA-Z]', ' ', text)` replaces non-English-letter characters with spaces.
- `lower()` normalizes capitalization, while `split()` produces word tokens.
- Stop-word removal reduces noise, but sentiment-bearing negations should be handled carefully.
- Stemming groups related word forms to reduce redundant vocabulary features.
- Flexible code should use the dataset's actual length rather than a hard-coded row count.

## Review questions

1. What does `[^a-zA-Z]` match?
2. Why are unwanted characters replaced with spaces instead of empty strings?
3. How does lowercasing affect the Bag-of-Words vocabulary?
4. Why might removing `not` hurt a sentiment-analysis model?
5. How does stemming differ from lemmatization?
6. Why is `range(len(dataset))` more reusable than `range(1000)`?
7. What data type does `review.split()` return?


---
# Study Notes: Completing the Text-Cleaning Pipeline

## Lesson objective

This lesson completes the preprocessing pipeline for the restaurant reviews. Each tokenized review is filtered to remove selected stop words, its remaining words are stemmed, and the tokens are joined back into a cleaned string. Every cleaned review is then added to `corpus`, ready for Bag-of-Words vectorization.

## 1. Recap of the earlier cleaning steps

Before this stage, each review passes through three transformations:

```python
review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])
review = review.lower()
review = review.split()
```

These statements:

1. replace non-English-letter characters with spaces;
2. normalize all letters to lowercase; and
3. split the string into a list of word tokens.

The review must be tokenized before word-level operations such as stop-word filtering and stemming can be applied.

## 2. Creating the Porter stemmer

```python
ps = PorterStemmer()
```

`PorterStemmer()` creates an object whose `stem()` method reduces words to normalized stems. It does not require constructor arguments.

Examples may include:

```python
ps.stem('loved')   # 'love'
ps.stem('loving')  # 'love'
```

Stemming can reduce redundant vocabulary features, but the resulting stem is not always a valid dictionary word. It is best understood as rule-based normalization rather than grammatical conversion.

> Efficiency note: the stemmer does not need to be recreated for every review. In reusable code, instantiate `PorterStemmer` once before the loop.

## 3. Preparing the stop-word set

For sentiment analysis, the notebook keeps the word `not` because it can reverse the meaning of an opinion:

```python
english_stopwords = set(stopwords.words('english')) - {'not'}
```

A set is suitable here because membership checks such as `word not in english_stopwords` are typically fast. As with the stemmer, construct this set once outside the review loop rather than rebuilding it for every word or document.

Why preserve negation?

- `food was good` expresses positive sentiment;
- `food was not good` usually expresses negative sentiment.

Removing `not` would make these cleaned reviews much harder for a simple model to distinguish. Other negations, including `no` and `never`, may also deserve task-specific treatment.

## 4. Filtering and stemming with a list comprehension

The central transformation can be written as:

```python
review = [
    ps.stem(word)
    for word in review
    if word not in english_stopwords
]
```

This is a **list comprehension**, a compact way to build a new list from an existing iterable. It is equivalent to:

```python
stemmed_words = []

for word in review:
    if word not in english_stopwords:
        stemmed_words.append(ps.stem(word))

review = stemmed_words
```

### Reading the comprehension

Although the expression appears before the loop in Python syntax, it is easiest to understand in this order:

1. `for word in review` — visit each token;
2. `if word not in english_stopwords` — retain the token only if it is not excluded;
3. `ps.stem(word)` — stem each retained token;
4. collect the results in a new list.

Stop words are therefore omitted from the cleaned review; they are not merely skipped during stemming and retained unchanged.

## 5. Joining the tokens back into a string

After filtering and stemming, `review` is still a list of tokens. The vectorizer expects each document as a string, so the words are joined with spaces:

```python
review = ' '.join(review)
```

The string before `.join()` is the separator. Using `' '` inserts one space between adjacent tokens.

```python
['love', 'restaurant']
# becomes
'love restaurant'
```

By contrast, `''.join(review)` would produce `loverestaurant`, incorrectly merging separate words.

## 6. Appending the cleaned review to the corpus

```python
corpus.append(review)
```

`append()` adds the completed review as one element of the corpus list. After all rows have been processed:

- `len(corpus)` should equal the number of reviews in the dataset;
- each corpus element should be a cleaned string; and
- the order should still correspond to the original dataset rows and sentiment labels.

Maintaining this order is essential because the feature row created from `corpus[i]` must match the target label stored in `y[i]`.

## 7. Complete, reusable cleaning loop

A polished version of the lesson's pipeline is:

```python
import re
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

ps = PorterStemmer()
english_stopwords = set(stopwords.words('english')) - {'not'}
corpus = []

for raw_review in dataset['Review']:
    review = re.sub('[^a-zA-Z]', ' ', raw_review)
    review = review.lower().split()
    review = [
        ps.stem(word)
        for word in review
        if word not in english_stopwords
    ]
    corpus.append(' '.join(review))
```

This form avoids hard-coding 1,000 rows and creates the reusable stemmer and stop-word set only once.

## 8. Validating the result

Before vectorization, inspect the processed corpus:

```python
print(len(corpus))
print(corpus[:5])
assert len(corpus) == len(dataset)
assert all(isinstance(review, str) for review in corpus)
```

These checks verify the number and type of cleaned documents. Comparing several original and processed reviews is also useful:

```python
for original, cleaned in zip(dataset['Review'].head(), corpus[:5]):
    print('Original:', original)
    print('Cleaned: ', cleaned)
    print()
```

## 9. Connection to the Bag-of-Words model

The completed corpus is the input to a text vectorizer. A Bag-of-Words representation creates one row per review and one feature column per selected vocabulary term. Each cell records information such as how often a term occurs in a review.

`raw reviews` → `cleaned corpus` → `vectorizer` → `feature matrix X`

The label column is extracted separately as `y`. A classifier can then learn the relationship between the numerical text features and positive or negative sentiment.

> Vectorizers commonly produce a sparse matrix because most reviews contain only a small fraction of the full vocabulary. The existing notebook later calls `.toarray()`, which converts that sparse representation into a dense NumPy array for the selected model. These are related but distinct representations.

## 10. Important limitations

- Stemming may merge words too aggressively or produce stems that are not real words.
- Removing stop words is not automatically beneficial; its effect should be tested.
- A unigram Bag-of-Words model largely ignores word order, so preserving `not` helps but does not fully encode phrases such as `not good`.
- The ASCII-only regular expression removes digits, emoji, apostrophes, and non-English letters, some of which may carry sentiment.
- Text preprocessing choices should be learned or fixed using training data only when they involve fitted components, preventing information leakage from the test set.

## Common pitfalls

- Writing the comprehension in the wrong order; Python expects `expression for item in iterable if condition`.
- Rebuilding `set(stopwords.words('english'))` for every word, which adds unnecessary work.
- Removing all negations without checking their importance to sentiment.
- Using `review.append(...)` when the cleaned document should be added to `corpus`.
- Forgetting the space in `' '.join(review)`.
- Allowing the number or order of corpus items to differ from the labels.

## Key takeaways

- `PorterStemmer().stem(word)` normalizes a token to a stem.
- A list comprehension can filter stop words and stem retained tokens in one expression.
- Sentiment-bearing negations such as `not` often require special treatment.
- `' '.join(tokens)` reconstructs a space-separated document string.
- `corpus.append(review)` preserves one cleaned document per dataset row.
- The corpus becomes the input used to construct the numerical feature matrix.

## Review questions

1. Why should the Porter stemmer usually be created outside the review loop?
2. In what order should the filtering-and-stemming list comprehension be interpreted?
3. What happens to a token when it is found in the stop-word set?
4. Why might `not` be removed from the English stop-word set?
5. What is the difference between `' '.join(words)` and `''.join(words)`?
6. Why must the order of documents in `corpus` match the order of labels in `y`?
7. Why is the Bag-of-Words output usually stored sparsely?


---
# Study Notes: Creating the Bag-of-Words Model

## Lesson objective

After cleaning the restaurant reviews, the next step is to convert the text into numerical features. A **Bag-of-Words (BoW)** model represents each review by the vocabulary terms it contains. The resulting document-term matrix can then be used as the feature matrix `X` for a machine-learning classifier.

## 1. Inspecting the cleaned corpus

Before vectorization, inspect a few cleaned reviews:

```python
print(corpus[:10])
```

For example, an original review such as:

```text
Wow... Loved this place.
```

may become approximately:

```text
wow love place
```

The punctuation and capitalization have been normalized, selected stop words have been removed, and `loved` has been stemmed to `love`.

## 2. Preserving negation

The sentence `Crust is not good` illustrates why preprocessing decisions must reflect the task. If `not` is removed as a stop word, the cleaned result becomes `crust good`, which can incorrectly appear positive.

One way to retain it is:

```python
all_stopwords = set(stopwords.words('english'))
all_stopwords.discard('not')
```

The course may use `remove('not')`, but `discard('not')` is slightly more defensive: `remove()` raises `KeyError` if the word is absent, whereas `discard()` does nothing in that case. A concise alternative is:

```python
all_stopwords = set(stopwords.words('english')) - {'not'}
```

Preserving `not` improves the input representation, but a unigram Bag-of-Words model still does not explicitly encode the phrase `not good`. Including bigrams is one possible extension.

## 3. What a Bag-of-Words matrix represents

A Bag-of-Words model creates a matrix in which:

- each **row** represents one document or review;
- each **column** represents one vocabulary feature; and
- each **cell** records the value of that term for that document.

For a count-based model, the values are term frequencies:

| Review | `food` | `good` | `not` | `service` |
|---|---:|---:|---:|---:|
| `food good good` | 1 | 2 | 0 | 0 |
| `service not good` | 0 | 1 | 1 | 1 |

The representation is called a **bag** of words because it records vocabulary occurrences while largely ignoring grammar and word order.

> Important correction: `CountVectorizer` uses integer counts by default. Its cells are not restricted to `0` and `1`; a repeated word can produce `2`, `3`, or another count. To create presence/absence features, set `binary=True`.

## 4. Tokenization and vocabulary construction

In this context, **tokenization** means splitting text into tokens. `CountVectorizer` performs tokenization and then builds a vocabulary that maps each retained token to a feature-column index. Vocabulary construction and count-matrix creation are related steps, but they are not both synonymous with tokenization.

The default word analyzer generally extracts alphanumeric tokens containing at least two characters. This means one-character tokens are ignored unless `token_pattern` or the tokenizer is changed.

## 5. Importing and creating `CountVectorizer`

```python
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(max_features=1500)
```

`CountVectorizer` converts a collection of text documents into a matrix of token counts. The name `cv` is only a local variable; a more descriptive name such as `vectorizer` is also common.

### `max_features`

When `max_features=1500`, the vocabulary is limited to at most 1,500 terms, ordered by total term frequency across the corpus. Without this argument, all terms that satisfy the other vectorizer rules are eligible.

This parameter controls dimensionality, but it does **not** guarantee that every retained word is relevant to sentiment or that every discarded word is irrelevant. Frequency is not the same as predictive value. The value should ideally be treated as a hyperparameter and evaluated using validation data.

Other useful vocabulary controls include:

- `min_df`: ignore terms appearing in fewer than a specified number or proportion of documents;
- `max_df`: ignore terms appearing in more than a specified number or proportion of documents;
- `ngram_range`: include unigrams, bigrams, or other n-grams;
- `binary=True`: store presence or absence rather than raw counts.

## 6. Fitting and transforming the corpus

The notebook uses:

```python
X = cv.fit_transform(corpus).toarray()
```

This performs two operations:

1. **Fit:** learn the vocabulary and assign each retained term a feature index.
2. **Transform:** count those vocabulary terms in every review and return the document-term matrix.

`fit_transform()` is an efficient combined operation; it does not literally “fit the corpus to `X`.” It learns the vectorizer's vocabulary from the documents and returns their numerical representation.

### Sparse versus dense output

`cv.fit_transform(corpus)` returns a SciPy sparse matrix. Most entries are zero because an individual review uses only a small fraction of the full vocabulary. Sparse storage records nonzero values efficiently.

Calling `.toarray()` converts that result into a dense NumPy array. A dense array is not inherently required for every feature matrix or classifier; many scikit-learn estimators accept sparse matrices directly. In this notebook, conversion is needed because `GaussianNB` expects dense input. Dense conversion can use substantial memory on larger corpora.

```python
X_sparse = cv.fit_transform(corpus)
X_dense = X_sparse.toarray()
```

## 7. Inspecting matrix dimensions and vocabulary

Use `.shape` to inspect the matrix:

```python
X.shape
# (number_of_reviews, number_of_features)

X.shape[0]  # number of reviews
X.shape[1]  # number of vocabulary features
```

This is clearer and works for both dense and sparse matrices. `len(X[0])` works after conversion to a dense array but is not the preferred general approach.

To see which word corresponds to each column:

```python
feature_names = cv.get_feature_names_out()
print(feature_names[:20])
print(len(feature_names))
```

The index of each item in `feature_names` matches its column index in `X`. The fitted mapping is also available through `cv.vocabulary_`.

## 8. Creating the target vector

The dependent variable is already numerical, so it can be extracted directly from the final dataset column:

```python
y = dataset.iloc[:, -1].values
```

Here:

- `:` selects every row;
- `-1` selects the last column; and
- `.values` returns the labels as a NumPy array.

For this dataset, `y` contains `0` for negative reviews and `1` for positive reviews. The following relationship should hold:

```python
assert X.shape[0] == y.shape[0]
```

## 9. Avoiding vocabulary leakage

For a rigorous evaluation, the vectorizer should learn its vocabulary from the training reviews only. Fitting it on the complete corpus before the train/test split allows information about the test-set vocabulary and term frequencies to influence feature selection, especially when `max_features`, `min_df`, or `max_df` is used.

A safer conceptual sequence is:

```python
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer

reviews_train, reviews_test, y_train, y_test = train_test_split(
    corpus,
    y,
    test_size=0.20,
    random_state=0
)

cv = CountVectorizer(max_features=1500)
X_train = cv.fit_transform(reviews_train)
X_test = cv.transform(reviews_test)
```

Notice that the test reviews use `transform()`, not `fit_transform()`. Unknown test-only words are ignored because the model must use the vocabulary learned from training data. A scikit-learn `Pipeline` is often the cleanest way to enforce this separation during validation and cross-validation.

## 10. Count features versus binary features

Choose the representation deliberately:

```python
# Raw occurrence counts
CountVectorizer(binary=False)  # default

# Presence/absence indicators
CountVectorizer(binary=True)
```

With count features, `good good` produces a value of `2` in the `good` column. With binary features, every nonzero count is set to `1`. Which representation works better depends on the data and classifier.

## 11. Pipeline overview

`cleaned corpus` → `learn training vocabulary` → `count terms` → `X_train and X_test` → `train classifier` → `predict` → `confusion matrix and accuracy`

Once `X` and `y` are available, the remaining steps follow a standard supervised-classification workflow: split the data, train the model, predict unseen labels, and evaluate its performance.

## Common pitfalls

- Assuming `CountVectorizer` creates only zeros and ones without setting `binary=True`.
- Treating vocabulary construction as identical to tokenization.
- Converting a large sparse matrix to a dense array without considering memory usage.
- Assuming the most frequent words are automatically the most predictive.
- Using `len(X[0])` instead of the clearer `X.shape[1]`.
- Fitting the vectorizer on the test data or the complete dataset before evaluation.
- Calling `fit_transform()` separately on the training and test sets, which creates incompatible feature columns.
- Forgetting that the cleaned corpus and target vector must remain row-aligned.

## Key takeaways

- Bag of Words converts documents into numerical vocabulary features.
- `CountVectorizer` performs tokenization, vocabulary construction, and occurrence counting.
- Counts are the default; use `binary=True` for zero-or-one indicators.
- `max_features=1500` retains at most 1,500 terms based on corpus-wide term frequency.
- `fit_transform()` learns the vocabulary and returns the document-term matrix.
- The matrix is sparse until `.toarray()` converts it to a dense NumPy array.
- Fit the vectorizer only on training text for unbiased model evaluation.
- Use `get_feature_names_out()` and `.shape` to inspect the fitted representation.

## Review questions

1. What do the rows, columns, and cells of a Bag-of-Words matrix represent?
2. Why can a default `CountVectorizer` cell contain a value greater than `1`?
3. Which parameter converts counts into presence/absence indicators?
4. What does `max_features=1500` actually select?
5. What is the difference between `fit_transform()` and `transform()`?
6. Why is a document-term matrix usually stored sparsely?
7. When can `.toarray()` become problematic?
8. How can you retrieve the vocabulary terms in column order?
9. Why should the vectorizer be fitted only on training reviews?
10. Why does preserving `not` help but not completely solve negation handling?


---
# Study Notes: Training and Evaluating the Sentiment Classifier

## Lesson objective

With the numerical feature matrix `X` and sentiment labels `y` prepared, the final stage is to split the observations, train a classifier, make predictions on unseen test reviews, and evaluate the results. The notebook uses Gaussian Naive Bayes and obtains an accuracy of `0.73` on one fixed test split.

## 1. Splitting the dataset

```python
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=0
)
```

This divides the feature rows and their corresponding labels into two aligned subsets:

- the **training set**, used to fit the model; and
- the **test set**, used only for final evaluation.

Because this dataset contains 1,000 observations, `test_size=0.20` assigns approximately 800 reviews to training and 200 to testing. `random_state=0` makes the shuffled split reproducible. It does not improve the model; it ensures that repeated runs use the same partition.

For classification, stratification is often a useful addition:

```python
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=0,
    stratify=y
)
```

`stratify=y` approximately preserves the class proportions in both subsets. This is especially important when the target classes are imbalanced.

> Evaluation reminder: in a rigorous workflow, split the review texts before fitting `CountVectorizer`. Fit the vocabulary on the training text and use only `transform()` on the test text. Splitting an already vectorized full corpus can leak test-set vocabulary information into training.

## 2. Training Gaussian Naive Bayes

```python
from sklearn.naive_bayes import GaussianNB

classifier = GaussianNB()
classifier.fit(X_train, y_train)
```

`fit()` estimates the class prior probabilities and the feature statistics required by the model. The fitted classifier can then assign a sentiment label to new feature rows.

Naive Bayes models make a strong conditional-independence assumption: given the class, they treat the features as independent. This assumption is rarely exactly true for language, but the models can still provide useful text-classification baselines.

### Choosing the Naive Bayes variant

The notebook uses `GaussianNB`, which assumes that each feature follows a Gaussian distribution within each class. Bag-of-Words features are nonnegative discrete counts, so that assumption is not a natural match.

`MultinomialNB` is specifically suited to discrete features such as word counts and is an important comparison model:

```python
from sklearn.naive_bayes import MultinomialNB

classifier = MultinomialNB()
classifier.fit(X_train, y_train)
```

`ComplementNB` is another text-oriented variant designed to improve multinomial Naive Bayes on imbalanced datasets. Other useful baselines include logistic regression and linear support vector classification. Model selection should be based on validation or cross-validation results rather than reputation alone.

## 3. Predicting the test labels

```python
y_pred = classifier.predict(X_test)
```

`predict()` returns one predicted class label for each row of `X_test`. The arrays must remain aligned:

```python
assert len(y_pred) == len(y_test)
```

The predictions and true labels can be displayed side by side:

```python
comparison = np.concatenate(
    (y_pred.reshape(-1, 1), y_test.reshape(-1, 1)),
    axis=1
)
print(comparison)
```

Each displayed row has the form `[predicted_label, true_label]`. This is useful for spot checks, but aggregate metrics provide a more reliable summary than manually scanning 200 predictions.

## 4. Creating the confusion matrix

```python
from sklearn.metrics import confusion_matrix, accuracy_score

cm = confusion_matrix(y_test, y_pred)
print(cm)
```

For binary labels ordered as `[0, 1]`, scikit-learn defines the matrix as:

```text
                 Predicted 0    Predicted 1
Actual 0              TN             FP
Actual 1              FN             TP
```

The notebook produces:

```text
[[55 42]
 [12 91]]
```

Its entries mean:

- **55 true negatives (TN):** negative reviews correctly predicted as negative;
- **42 false positives (FP):** negative reviews incorrectly predicted as positive;
- **12 false negatives (FN):** positive reviews incorrectly predicted as negative;
- **91 true positives (TP):** positive reviews correctly predicted as positive.

A common mistake is to describe `42` merely as “incorrect positive predictions” or `12` as “incorrect negative predictions” without stating the actual and predicted classes. The terms false positive and false negative remove that ambiguity.

## 5. Computing and interpreting accuracy

```python
accuracy = accuracy_score(y_test, y_pred)
print(accuracy)
# 0.73
```

Accuracy is the fraction of all observations classified correctly:

```text
accuracy = (TN + TP) / (TN + FP + FN + TP)
         = (55 + 91) / 200
         = 0.73
```

Thus, the classifier predicts 146 of the 200 test reviews correctly on this particular split. The test set contains 97 negative and 103 positive reviews, so the majority-class baseline is `103 / 200 = 51.5%`. The observed 73% is better than that simple baseline, but one split alone does not establish how reliably the model generalizes.

## 6. Metrics beyond accuracy

Accuracy can hide important differences between error types. For the positive class in this result:

```text
precision = TP / (TP + FP) = 91 / 133 ≈ 0.684
recall    = TP / (TP + FN) = 91 / 103 ≈ 0.883
F1-score  = harmonic mean of precision and recall ≈ 0.771
```

The high positive-class recall means that most truly positive reviews are detected. The lower precision reflects the 42 negative reviews incorrectly labeled positive. Negative-class recall, also called specificity, is `55 / 97 ≈ 0.567`.

Generate a fuller report with:

```python
from sklearn.metrics import classification_report

print(classification_report(
    y_test,
    y_pred,
    target_names=['negative', 'positive']
))
```

Which metric matters most depends on the application and the relative cost of false positives and false negatives.

## 7. Understanding what the model learned

The classifier has not learned to understand English in the human sense. It has learned statistical associations between numerical word features and sentiment labels in this dataset. Because Bag of Words largely ignores word order, context, sarcasm, and compositional meaning, its language representation is limited.

For example, preserving `not` is helpful, but unigram features still represent `not good` as two independent terms. Using `ngram_range=(1, 2)` can add bigrams such as `not good`, although this increases the feature space and must be evaluated carefully.

## 8. Handling contractions correctly

Simply excluding `isn't` from a stop-word list may not work with the current preprocessing order. The regular expression removes the apostrophe first, turning `isn't` into tokens similar to `isn` and `t`. The original contraction no longer exists when stop-word filtering occurs.

A better approach is to normalize contractions before removing punctuation:

```text
isn't → is not
wasn't → was not
didn't → did not
```

This preserves the negation token explicitly. Contraction handling should cover the desired vocabulary systematically rather than patching only one example.

## 9. Improving the experiment responsibly

Possible comparisons include:

- `MultinomialNB` or `ComplementNB`;
- logistic regression or a linear support vector classifier;
- count features versus `binary=True`;
- unigram versus unigram-and-bigram features;
- different `max_features`, `min_df`, and `max_df` values;
- stemming versus lemmatization or no stemming;
- CountVectorizer versus TF-IDF features;
- improved contraction and negation handling.

Do not repeatedly choose changes based on the same test-set score. That turns the test set into a tuning set and makes the final estimate optimistic. Use a validation set or cross-validation for model selection, then evaluate the chosen configuration once on an untouched test set.

A scikit-learn `Pipeline` can keep vectorization and classification together so that cross-validation fits all learned preprocessing only on each training fold.

## 10. Reproducible evaluation checklist

1. Keep labels aligned with their reviews.
2. Reserve an untouched test set.
3. Fit the vectorizer only on training text.
4. Use stratification when appropriate.
5. Establish a simple baseline.
6. compare models using validation data or cross-validation;
7. report the confusion matrix and class-specific metrics; and
8. evaluate the final selected approach once on the test set.

## Common pitfalls

- Calling the algorithm “Naive Base” instead of **Naive Bayes**.
- Treating `random_state` as a model-performance parameter rather than a reproducibility control.
- Omitting stratification when class proportions matter.
- Fitting text-vectorization rules on the complete dataset before splitting.
- Assuming Gaussian Naive Bayes is automatically the best Naive Bayes model for word counts.
- Reversing the arguments to metrics; use `metric(y_true, y_pred)`.
- Misreading rows as predicted classes and columns as actual classes in scikit-learn's confusion matrix.
- Reporting only accuracy without comparing a baseline or examining class-specific errors.
- Repeatedly tuning against the test set.
- Claiming that Bag of Words gives the model human-like language understanding.

## Key takeaways

- `train_test_split` creates aligned training and test subsets; `random_state` makes the split reproducible.
- `stratify=y` helps preserve class proportions.
- `GaussianNB.fit()` trains the course model, but `MultinomialNB` is often a more natural baseline for word counts.
- `predict()` returns one label per test observation.
- For `[[55, 42], [12, 91]]`, the entries are TN=55, FP=42, FN=12, and TP=91.
- The 73% accuracy means 146 of 200 test predictions were correct.
- Precision, recall, F1-score, and the confusion matrix reveal behavior that accuracy alone cannot show.
- Model selection belongs on validation data; the test set should remain untouched until final evaluation.

## Review questions

1. What do `test_size=0.20` and `random_state=0` control?
2. Why might `stratify=y` be useful?
3. Why is `MultinomialNB` often a better conceptual match for word counts than `GaussianNB`?
4. In what order should arguments be passed to `confusion_matrix` and `accuracy_score`?
5. What do the four entries in `[[55, 42], [12, 91]]` represent?
6. How is the 73% accuracy calculated?
7. What do precision and recall reveal that accuracy may hide?
8. Why is excluding the literal string `isn't` insufficient after punctuation removal?
9. Why should model improvements be selected using validation data rather than the test set?
10. In what sense does a Bag-of-Words classifier fall short of understanding English?
